src/sentry/integrations/pipeline.py

```diff
@@ -78,17 +80,30 @@ def _finish_pipeline(self, data):
             if created:
                 idp.update(config=identity_config)
 
-            Identity.objects.get_or_create(
-                idp=idp,
-                user=self.request.user,
-                external_id=identity['external_id'],
-                defaults={
-                    'status': IdentityStatus.VALID,
-                    'scopes': identity['scopes'],
-                    'data': identity['data'],
-                    'date_verified': timezone.now(),
-                },
-            )
+            identity_data = {
+                'status': IdentityStatus.VALID,
+                'scopes': identity['scopes'],
+                'data': identity['data'],
+                'date_verified': timezone.now(),
+            }
+
+            try:
+                Identity.objects.get_or_create(
+                    idp=idp,
+                    user=self.request.user,
+                    external_id=identity['external_id'],
+                    defaults=identity_data,
+                )
+            except IntegrityError:
+                # If the external_id is already used for a different user or
+                # the user already has a different external_id, reparent it
+                lookup = Q(external_id=identity['external_id']) | Q(user=self.request.user)
+
+                Identity.objects.filter(lookup, idp=idp).update(
```

Comment:
will this handle the case where there are 2 objects? similar to https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/slack/link_identity.py#L99

Reply:
Maybe what will be easier is to just delete the identities then create a new one

Reply:
It should now f47cb90